### PR TITLE
Speed up network list action

### PR DESCRIPTION
**Pull request by xianlubird**
**\- What I did**
Speed up docker network list 

**\- How I did it**
concurrent get network detail info

**\- How to verify it**
run docker network ls

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @unclejack @LK4D4 ptal

File: api/server/router/network/network_routes.go [hunk 1]
```diff
@@ -34,16 +35,29 @@ func (n *networkRouter) getNetworksList(ctx context.Context, w http.ResponseWrit
...
-		list = append(list, *n.buildNetworkResource(nw))
+		wg.Add(1)
+		go func(tmpNw libnetwork.Network) {
+			networkResourceChannel <- *n.buildNetworkResource(tmpNw)
+		}(nw)
 	}
 
+	go func() {
+		for resource := range networkResourceChannel {
+			list = append(list, resource)
```
Comment (unclejack) [inline api/server/router/network/network_routes.go]: Isn't this a race?

Comment (xianlubird) [inline api/server/router/network/network_routes.go]: `networkResourceChannel ` is a channel which buffer size is 1. Then the last `goroutine` will only set one value  into channel at the same time.  This routine will read this channel synchronously, it won't cause some race condition. @unclejack

**Comment by LK4D4** [comment]
@xianlubird Can you measure performance gain with time and prints and provide it here? Thanks!

**Comment by xianlubird** [comment]
root@node-1:/home/vagrant# docker network ls
    NETWORK ID          NAME                 DRIVER              SCOPE
    ce83d114a30f        bridge               bridge              local
    284fbe8ee93c        docker_gwbridge      bridge              local
    09f6fa4a3f3f        host                 host                local
    f1f0b403498b        multi-host-network1   overlay             global
    fea0480c6039        multi-host-network2   overlay            global

I run this command `docker network ls` in a machine which have 2 core and 4G memory with 110 containers. With this PR, this commander will cost about 12 seconds to return. After using this PR, it will return at about 5 seconds.

@LK4D4 @unclejack @thaJeztah @LK4D4

**Comment by allencloud** [comment]
I was wondering if we can call the `Info()` via libnetwork to k/v in batch? since this pr did not decrease the call numbers.
Have we discussed the root cause of the situation?

**Comment by xianlubird** [comment]
In my opinion, the method `Info()` should only return one network's info, rather than all the subnet info of this overlay.  If we want to get all endpoint of this overlay network, we'd better define a new method to this Interface of libnetwork. However, these will affect many other places. So I think concurrent read these info is the easiest way to reduce API's response time.
@allencloud WDYT ?

**Comment by xianlubird** [comment]
#27729

File: api/server/router/network/network_routes.go [hunk 2]
```diff
@@ -34,16 +35,30 @@ func (n *networkRouter) getNetworksList(ctx context.Context, w http.ResponseWrit
...
 		}
-		list = append(list, *n.buildNetworkResource(nw))
+		wg.Add(1)
+		go func(tmpNw libnetwork.Network) {
+			tmpNwr := *n.buildNetworkResource(tmpNw)
+			networkResourceChannel <- tmpNwr
+		}(nw)
 	}
 
+	go func() {
```
Comment (cpuguy83) [inline api/server/router/network/network_routes.go]: This goroutine will run forever

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Comment by allencloud** [comment]
Need a rebase, @xianlubird

**Comment by thaJeztah** [comment]
@xianlubird I see something went wrong with your rebase; will you be reopening a PR to replace this?

**Comment by xianlubird** [comment]
@thaJeztah I won't reopen one to replace this. I think the method of concurrent get network info isn't a good way to speed up network list action. We should add a batch method for k/v interface. So I close this PR.

**Pull request closed without merging**